### ISSUE TITLE
Add proper header claims in SD-JWT

### ISSUE THREAD

**Issue by peppelinux** [oauth-wg/oauth-selective-disclosure-jwt#60]
We have 

````
{
  "alg": "RS256"
}
````

we may have something like

````
{
  "typ": "SD-JWT"
  "alg": "RS256"
}
````

**Comment by tlodderstedt** [comment]
+1

**Comment by peppelinux** [comment]
I suppose also to have at least `kid` or `x5t` or `jwk` or `x5c` in the headers.

**Comment by peppelinux** [comment]
and also the typ header claim "SD-JWT-R"

do we have nay prefernces on the name of these typ(s)?

"selective-disclosure-jwt+r" and things or the KISS approach? (I'm for the second one)

**Comment by danielfett** [comment]
If I look at what we have chosen for DPoP, I think we should go for `sd+jwt` and `sd-r+jwt` or similar. @b---c Do you have an opinion on this?

**Comment by bc-pi** [comment]
Agree with `sd+jwt`. 

Not sure about the `sd-r+jwt` because the SD JWT release is more stuff than a single JWT. And typ is describing the JWT containing it.

**Comment by peppelinux** [comment]
@bc-pi

This is the SD-JWT-R taken from draft and demo code
````
eyJhbGciOiAiUlMyNTYifQ.eyJub25jZSI6ICIyR0xDNDJzS1F2ZUNmR2ZyeU5STjl3Iiw
gImF1ZCI6ICJodHRwczovL2V4YW1wbGUuY29tL3ZlcmlmaWVyIiwgIl9zZCI6IHsiZ2l2Z
W5fbmFtZSI6ICJbXCI2SWo3dE0tYTVpVlBHYm9TNXRtdlZBXCIsIFwiSm9oblwiXSIsICJ
mYW1pbHlfbmFtZSI6ICJbXCJlSThaV205UW5LUHBOUGVOZW5IZGhRXCIsIFwiRG9lXCJdI
iwgImFkZHJlc3MiOiAiW1wiUGMzM0pNMkxjaGNVX2xIZ2d2X3VmUVwiLCB7XCJzdHJlZXR
fYWRkcmVzc1wiOiBcIjEyMyBNYWluIFN0XCIsIFwibG9jYWxpdHlcIjogXCJBbnl0b3duX
CIsIFwicmVnaW9uXCI6IFwiQW55c3RhdGVcIiwgXCJjb3VudHJ5XCI6IFwiVVNcIn1dIn1
9.b0hG3v71rzHvtoDTdroZ9m-lt9tf8nobFKb2YGiyGOjIklfcKc2KWj72oi_tBKcOCqZh
dX6IV4BRXIw-aspQfLh-xBrNLuGqiC-Y3rZBBlWw0WWnbbtsy1tj8yZOiXBr8vO6mCgZGA
d4MgPYPd-QzOr9ukObYDRB4I24xHrqlAEYPJIzSw9MI_dEmIkNnAuIfLQKiuyTqVVVp6Ly
pBIz6fBLm6NOLC4-uVXlOzI91iT4zlkrhP0-vj8TmfB-XL9aD3-xqytvLBHTESct49OSRZ
FrwkLUKTM56_6KW3pG7Ucuv8VnpHXHIka0SGRaOh8x6v5-rCQJl_IbM8wb7CSHvQ
````

it's a signed JWT with the structure above

````
{
  "alg": "RS256"
}
.
{
  "nonce": "2GLC42sKQveCfGfryNRN9w",
  "aud": "https://example.com/verifier",
  "_sd": {
    "given_name": "[\"6Ij7tM-a5iVPGboS5tmvVA\", \"John\"]",
    "family_name": "[\"eI8ZWm9QnKPpNPeNenHdhQ\", \"Doe\"]",
    "address": "[\"Pc33JM2LchcU_lHggv_ufQ\", {\"street_address\": \"123 Main St\", \"locality\": \"Anytown\", \"region\": \"Anystate\", \"country\": \"US\"}]"
  }
}
````

So if we agree we can add `sd-r+jwt` in the SD-JWT-R

**Comment by peppelinux** [comment]
PR here
https://github.com/oauthstuff/draft-selective-disclosure-jwt/pull/83

Code here
https://github.com/oauthstuff/draft-selective-disclosure-jwt/pull/61/commits/d49b8ffe82392b3a4948393863e203ad25885b4a
https://github.com/oauthstuff/draft-selective-disclosure-jwt/pull/61/commits/4a4575f36906c61471179b8b530f47d980853ba1

**Comment by bc-pi** [comment]
My previous comment had some confusion about what exactly is SD JWT release. I thought (from misunderstanding or maybe text from a previous revision of the draft) that the SD-JWT-R was a term for the whole package.

**Comment by bc-pi** [comment]
Thinking about this a bit more and I don't think that `typ` value(s) should be defined or mandated by this draft. 

[RFC 8725 - JSON Web Token Best Current Practices](https://tools.ietf.org/html/rfc8725.html) talks about using explicit typing as a security mechanism to prevent "Cross-JWT Confusion" but typing something as an SD-JWT doesn't accomplish the same thing and might preclude use in contexts where their is a typ already in use (i.e. `at+jwt` for an access token perhttps://datatracker.ietf.org/doc/html/rfc9068#section-2.1).

**Comment by bc-pi** [comment]
It's subtle but this isn't exactly a type of JWT and the title 'Selective Disclosure JWT' is potentially misleading. The document is more defining some structures/claims for use in JWT that can (help) accomplish selective disclosure.  'Selective Disclosure with JWT' or 'Selective Disclosure for JWT' might be more appropriate.

**Comment by danielfett** [comment]
Coming back to this after a looong time...

What about this solution?
- Add a `typ` header, e.g., `sd+jwt` to the SD-JWT.
- Define a new header, `_sd_typ`, that defines *what the SD-JWT is supposed to represent*, e.g., `at`, `secevent` or similar.
- Define that the verifier must replace `typ` with the contents of `_sd_typ` before passing the data to the application.

**Comment by bc-pi** [comment]
> What about this solution? 

I usually like your ideas @danielfett but would be strongly opposed to that. 

IMHO this draft should not say or mandate anything about "typ" headers. Leave that to the application to use (or not) as appropriate. But `"typ": "sd+jwt"` isn't really meaningful or useful on it's own (similar to `"typ": "JWT"`, the WTF from RFC7519) and so we don't need special SD-JWT features for it. 

It _maybe_ worthwhile to define media type(s) for the combined format(s) that could be used to signal what the thing is in like HTTP `Content-Type` or `cty` header of an enclosing JWE.  Maybe.

**Comment by Sakurann** [comment]
Resolved with PR #249
